highfidelity/hifi
# Android - Daydream integration

**Test plan**

Mode switching
- Test the app allows enter and exit VR mode
- The app should remember last mode the user has been in
- Entering VR mode should be always preceded by the daydream transition that ask the user to place the phone into the headset
- If the app needs to perform some operation in 2D while being in VR mode it must show the daydream transition that ask the user to remove the phone from the headset. i.e. starting from Daydream App when some permission is not granted

Controller
Info about the controller  >  https://support.google.com/daydream/answer/7184597?hl=en
- Clicking the touchpad left and right must rotate my avatar 
- Clicking the touchpad down and up must move/walk back and forward
- Clicking the center of the touchpad can be used to drag objects
- Use the app button to teleport
- Hold daydream button to re-center, in case you've not started pointing out forward

Head tracking
- Head rotation must be tracked and reflected in my avatar

---

How to use:
- In this PR (temporarily) the "MY VIEW" button switches to VR mode. Can't restore afterwards, see "Pending features and tasks".

Pending features and tasks:
- Upload gvr library to use in precompiled folder (with current version, we should copy by hand `gvr-android-sdk-1.170` into hifi_externals/precompiled folder) - DONE
- Correct eyes _focus_ or fov (maybe stop using rotate and interpret head pose correctly) - DONE
- Bring back the controller - Currently we don't have the UI that asks the user to 'turn on' the controller, but behind scenes we still need the user to activate the controller pressing the Home Button for some seconds. We *need* to implement our own version of this kind of instructions as that's given by GvrLayout that we are not using. 
- UI for selecting daydream mode
- Way to exit daydream mode (now, the only way is to clear data after setting it up pressing 'View' ) - DONE close button
- Extra daydream UI: close and settings are not present in our view. We'd use GvrLayout or make a similar thing to simulate it, with a catch, it should be after the daydream or hifi stencils are _drawn_.  - DONE settings button (goes to hifi settings)
- This may crash when switching to VR by the following assert: `A/libc: /Users/cduarte/dev/workspace-hifi/hifinw/libraries/gpu/src/gpu/Buffer.cpp:102: void gpu::Buffer::Update::apply() const: assertion "buffer._applyUpdateCount == updateNumber" failed` We need to find out why's that. - STILL PENDING BUT ONLY HAPPENS ON DEV BUILDS
- Crash in devices with no support of (vertex) shader storage blocks: `The number of vertex shader storage blocks (1) is greater than the maximum number allowed (0)` -- Addressed in PR https://github.com/highfidelity/hifi/pull/14379 - DONE

DO NOT MERGE yet
